### PR TITLE
[#8949] lab external id

### LINKED ISSUE #8949: [DEMIS2SORMAS] Compare the list of Organization Id's with facilities in SORMAS [3]
### Feature Description

When https://github.com/hzi-braunschweig/sormas-demis-adapter/issues/49 gets implemented, not only one but a list of Id's can map to available facilities (specifically laboratories) in SORMAS.
### Problem Description

Thus far, only one Id is considered when mapping the value in `labMessage.labExternalId` to facilities in SORMAS. 
### Proposed Change

- [x] Rename the `LabMessageDto.labExternalId` to `.labExternalIds` [with an 's'] and enable the attribute to store multiple values (via storage as JSON string).
- [x] Consider all available Id's in the list of `labMessage.labExternalId` and compare them with existing facilities
    - [x] If no match was found, select `OTHER` in the facility field and write the respective facility name into the details field as before.
    - [x] If just one Id matches an existing facility, set that facility as reference.
    - [x] If more than one Id matches existing facilities, leave the field empty since the user has to make a decision about which facility is supposed to be referenced.
- [x] Implement the points above as well for test reports.
- [x] Cover the new functionalities with unit tests
- [x] Make sure that the changes don't break sharing lab messages (and test reports) via S2S.
### Additional Information

Related ticket in the SORMAS_DEMIS-adapter: [#49](https://github.com/hzi-braunschweig/sormas-demis-adapter/issues/49).

### PR DESCRIPTION
**Pull request by JonasCir**
Fixes #8949 
@FredrikSchaeferVitagroup ~My tests with regard to S2S are still pending, but this is still ready for review.~

### CODE REVIEW

File: sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java
Comment (vidi42) [inline sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java:35]: Should it? To my understanding a lab message could point to multiple facilities, but a facility is still identified by a single external id

File: sormas-api/src/test/java/de/symeda/sormas/api/utils/DataHelperTest.java
Comment (vidi42) [inline sormas-api/src/test/java/de/symeda/sormas/api/utils/DataHelperTest.java:137]: We should remove/resolve the `@Ignore` before merging

**Review by vidi42 — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java
Comment (JonasCir) [reply sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java:35]: Correct. I think we can currently leave it this way. I added a remark for reference DTOs to #9389 in case we want to change something there.

File: sormas-api/src/test/java/de/symeda/sormas/api/utils/DataHelperTest.java
Comment (JonasCir) [reply sormas-api/src/test/java/de/symeda/sormas/api/utils/DataHelperTest.java:137]: true

**Comment by JonasCir** [comment]
@vidi42 Is it ok to merge this even if the corresponding DEMIS adapter changes are pending?

**Comment by vidi42** [comment]
> @vidi42 Is it ok to merge this even if the corresponding DEMIS adapter changes are pending?

We should probably wait for the demise issue to be ready to be merged as well.

File: sormas-backend/src/main/resources/sql/sormas_schema.sql
Comment (FredrikSchaeferVitagroup) [inline sormas-backend/src/main/resources/sql/sormas_schema.sql:11608]: Those commands don't work (anymore), because the labmessage tables got renamed to externalmessage.

File: sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java
Comment (FredrikSchaeferVitagroup) [reply sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java:35]: I also think this should not be a list, especially not for #8949. Shouldn't this comment just be removed?

**Review by FredrikSchaeferVitagroup — CHANGES_REQUESTED** [review]

File: sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java
Comment (JonasCir) [reply sormas-api/src/main/java/de/symeda/sormas/api/InfrastructureDataReferenceDto.java:35]: removed

File: sormas-backend/src/main/resources/sql/sormas_schema.sql
Comment (JonasCir) [reply sormas-backend/src/main/resources/sql/sormas_schema.sql:11608]: fixed

File: sormas-api/src/test/java/de/symeda/sormas/api/utils/DataHelperTest.java
Comment (JonasCir) [reply sormas-api/src/test/java/de/symeda/sormas/api/utils/DataHelperTest.java:137]: removed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/ExternalMessageMapper.java
Comment (FredrikSchaeferVitagroup) [inline sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/ExternalMessageMapper.java:339]: After the db migration, I got something like `{NULL}` in the labexternalids column of the for one entry.
This seems to be a list with a null element. This if statement yields true in that scenario, which it probably shouldn't... Or?

**Review by FredrikSchaeferVitagroup — CHANGES_REQUESTED** [review]

File: sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/ExternalMessageMapper.java
Comment (FredrikSchaeferVitagroup) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/ExternalMessageMapper.java:339]: When trying to process the external message with the `{NULL}` entry, the mapping crashes.

File: sormas-api/src/main/java/de/symeda/sormas/api/externalmessage/ExternalMessageDto.java
Comment (JonasCir) [reply sormas-api/src/main/java/de/symeda/sormas/api/externalmessage/ExternalMessageDto.java:80]: fixed

File: sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/ExternalMessageMapper.java
Comment (JonasCir) [reply sormas-ui/src/main/java/de/symeda/sormas/ui/externalmessage/ExternalMessageMapper.java:340]: @FredrikSchaeferVitagroup can you try again, I use `.filter(Objects::nonNull)` which should remove the object

**Review by FredrikSchaeferVitagroup — APPROVED** [review]

**Pull request merged**
